components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java

```diff
@@ -3376,4 +3380,39 @@ public String getMediationNameFromConfig(String config) {
         }
         return null;
     }
+
+    /**
+     * validate user inout scopes
+     *
+     * @param api api information
+     * @throws APIManagementException throw if validation failure
+     */
+    private void validateScopes(API api) throws APIManagementException {
+        APIIdentifier apiId = api.getId();
+        for (Scope scope : api.getScopes()) {
+            if (!(APIUtil.isWhiteListedScope(scope.getName()))) {
+                String username = RestApiUtil.getLoggedInUsername();
+                String tenantDomain = RestApiUtil.getLoggedInUserTenantDomain();
+                int tenantId = APIUtil.getTenantIdFromTenantDomain(tenantDomain);
+                APIProvider apiProvider = RestApiUtil.getProvider(username);
+
+                if (apiProvider.isScopeKeyAssigned(apiId, scope.getName(), tenantId)) {
+                    RestApiUtil
+                            .handleBadRequest("Scope " + scope.getName() + " is already assigned by another API", log);
+                }
+            }
+            if (scope.getDescription() == null || scope.getDescription().trim().isEmpty()) {
+                RestApiUtil.handleBadRequest("Scope cannot have empty description", log);
+            }
+            if (scope.getRoles() != null) {
+                for (String aRole : scope.getRoles().split(",")) {
+                    boolean isValidRole = APIUtil.isRoleNameExist(apiId.getProviderName(), aRole);
```

Comment:
why we need to pass provider name to check role existents ? Is it due to get the tenant domain?

Reply:
yes. Existing APIUtil#isRoleNameExist() require tenant ID. it is extracted from username.